packages/network-controller/src/NetworkController.ts

```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+            : { ...newRpcEndpointFields, networkClientId: uuidV4() };
+        operations.push({
+          type: 'add',
+          value: newRpcEndpoint,
+        });
+      }
+    }
+
+    const newRpcEndpoints: RpcEndpoint[] = [];
+    for (const operation of operations) {
+      if (operation.type === 'remove') {
+        const networkClient = this.getNetworkClientById(
+          operation.value.networkClientId,
+        );
+        networkClient.destroy();
+        delete autoManagedNetworkClientRegistry[
+          networkClient.configuration.type
+        ][operation.value.networkClientId];
+      } else {
+        if (operation.type === 'add') {
+          if (operation.value.type === RpcEndpointType.Infura) {
+            autoManagedNetworkClientRegistry[NetworkClientType.Infura][
+              operation.value.networkClientId
+            ] = createAutoManagedNetworkClient({
+              type: NetworkClientType.Infura,
+              chainId: newChainId,
+              network: operation.value.networkClientId,
+              infuraProjectId: this.#infuraProjectId,
+              ticker: newNativeTokenName,
+            });
+          } else {
+            autoManagedNetworkClientRegistry[NetworkClientType.Custom][
+              operation.value.networkClientId
+            ] = createAutoManagedNetworkClient({
+              type: NetworkClientType.Custom,
+              chainId: newChainId,
+              rpcUrl: operation.value.url,
+              ticker: newNativeTokenName,
+            });
+          }
+        }
+        newRpcEndpoints.push(operation.value);
+      }
+    }
+
+    const updatedNetworkConfiguration = {
+      ...fields,
+      rpcEndpoints: newRpcEndpoints,
+    };
 
     this.update((state) => {
-      delete state.networkConfigurations[networkConfigurationId];
+      if (newChainId !== existingChainId) {
+        delete state.networkConfigurationsByChainId[existingChainId];
+      }
+
+      state.networkConfigurationsByChainId[newChainId] =
+        updatedNetworkConfiguration;
     });
 
-    const customNetworkClientRegistry =
-      autoManagedNetworkClientRegistry[NetworkClientType.Custom];
-    const existingAutoManagedNetworkClient =
-      customNetworkClientRegistry[networkClientId];
-    existingAutoManagedNetworkClient.destroy();
-    delete customNetworkClientRegistry[networkClientId];
+    this.#networkConfigurationsByNetworkClientId =
+      buildNetworkConfigurationsByNetworkClientId(
+        this.state.networkConfigurationsByChainId,
+      );
+
+    return updatedNetworkConfiguration;
+  }
+
+  /**
+   * Destroys and unregisters the network identified by the given chain ID, also
+   * removing the associated network configuration from state.
+   *
+   * @param chainId - The chain ID associated with an existing network.
+   * @throws if `chainId` does not refer to an existing network configuration,
+   * or if the currently selected network is being removed.
+   * @see NetworkConfiguration
```

Comment:
nit:
```suggestion
   * @see {@link NetworkConfiguration}
```

Reply:
Fixed in [a600e92](https://github.com/MetaMask/core/pull/4286/commits/a600e9244be9f77c88e9369f6aab7c833b0374a1).